onyo/lib/onyo.py

```diff
@@ -1240,3 +1241,128 @@ def unset(self, paths: Iterable[Union[Path, str]], values: str,
             self.restore()
 
         return diff
+
+    #
+    # GET
+    #
+    # TODO: This should be a staticmethod in utils
+    def _walk(self, path, depth) -> Generator:
+        """Recursively yield filepaths up to a given depth"""
+        depth -= 1
+        with os.scandir(path) as p:
+            for entry in p:
+                # ignore hidden files and folders
+                # TODO: Better to use _is_protected_path or ignore hidden?
+                #   `if self._is_protected_path(entry.path):`
+                if entry.name.startswith('.'):
+                    continue
+
+                # yield discovered assets
+                if not entry.is_dir():
+                    yield entry.path
+
+                # explore dirs if depth allows
+                if entry.is_dir() and depth >= 0:
+                    yield from self._walk(entry.path, depth)
+
+    @staticmethod
+    def _filter_by_name(assets: List[str], filters: dict) -> Generator:
+        """Filter assets by filename to prevent redundant asset loading"""
+        keys = ['type', 'make', 'model', 'serial']
+        for asset in assets:
+            split = set(zip(keys, re.split('[_.]', os.path.basename(asset))))
+            if set(filters.items()).issubset(split):
+                yield asset
+
+    @staticmethod
+    def _get_asset_data(
+            assets: Union[Generator[Any, Any, Any], List[Any]], keys: Set[str],
+            filters: Union[dict, None] = None) -> Generator:
+        """Return data from asset keys for assets matching filters"""
+        yaml = YAML(typ='safe')
+        default_keys = ['type', 'make', 'model', 'serial']
+        string_types = {  # TODO: Do we have a list of accepted string types?
+            '[int]': int, '[float]': float, '[str]': str, '[list]': list,
+            '[dict]': dict, '[bool]': bool}
+
+        for asset in assets:
+            try:
+                data = yaml.load(Path(asset))
+            except scanner.ScannerError as e:
+                log.error(e)
+                continue
+
+            # add type,make,model,serial to data if key does not exist
+            default_data = dict(zip(
+                default_keys, re.split('[_.]', os.path.basename(asset))))
+            data = default_data | data
+
+            # no filters
+            if not filters:
+                yield asset, {k: v for k, v in data.items() if k in keys}
+                continue
+
+            # there are filters targeting non-default keys but no data
+            if not data and all(v == '[unset]' for v in filters.values()):
```

Comment:
The `_get_asset_data` function yields all requested asset keys for which the assets match the filter(s). The (outer) `for` loop performs the simple filters, this being one of them. It checks if there both is no data inside of the asset file **and** _all_ of the filters are only searching for `key=[unset]`. If both of these are true, then we have a match!

It should have yielded the asset though, so I fixed that ;)

Reply:
I'm still left confused by this.

First, the comment:
```
# there are filters targeting non-default keys but no data
```
What is a `default` key?

```
if not data and all(v == '[unset]' for v in filters.values()):
```
So, if I understand it correctly what you're trying to do here: if a user wants to filter for assets that don't have a particular field set. Thus you check if there's no asset data and any key-value that's `<unset>`. If so, yield the asset.

It is unclear to me why this is written as a special case, rather than as the normal path of checking whether a key is unset in an asset (regardless if that asset is empty or not).

Also, you did not address my point about `all()` vs `any()`. As I read it, if there is no `data`, and there are no `filters`, then the asset will be yielded.

Reply:
This has been reworked as part of `Filter.match(asset)`. I still use the `default_key` terminology, which refers to the keys in the name of the asset (i.e., `laptop_apple_macbookpro.1` as `type_make_model.serial`). These are required by virtue of the naming convention. If you prefer, I can rename them. Maybe `required_keys`? Suggestions welcome!